presto-main/src/main/java/io/prestosql/operator/LookupJoinOperator.java

```diff
@@ -41,587 +44,685 @@
+            this.yieldSignal = requireNonNull(yieldSignal, "yieldSignal is null");
+            this.memoryTrackingContext = requireNonNull(memoryTrackingContext, "memoryTrackingContext is null");
+            this.spillContext = requireNonNull(spillContext, "spillContext is null");
+        }
 
-    private boolean tryFetchLookupSourceProvider()
-    {
-        if (lookupSourceProvider == null) {
-            if (!lookupSourceProviderFuture.isDone()) {
-                return false;
+        public ListenableFuture<?> isBlocked()
+        {
+            if (unspilledLookupSource.isPresent()) {
+                // Unspilling can happen only after lookupSourceProviderFuture was done.
+                return unspilledLookupSource.get();
             }
-            lookupSourceProvider = requireNonNull(getDone(lookupSourceProviderFuture));
-            statisticsCounter.updateLookupSourcePositions(lookupSourceProvider.withLease(lookupSourceLease -> lookupSourceLease.getLookupSource().getJoinPositionCount()));
-        }
-        return true;
-    }
 
-    private Page spillAndMaskSpilledPositions(Page page, IntPredicate spillMask)
-    {
-        checkState(spillInProgress.isDone(), "Previous spill still in progress");
-        checkSuccess(spillInProgress, "spilling failed");
-
-        if (!spiller.isPresent()) {
-            spiller = Optional.of(partitioningSpillerFactory.create(
-                    probeTypes,
-                    getPartitionGenerator(),
-                    operatorContext.getSpillContext().newLocalSpillContext(),
-                    operatorContext.newAggregateSystemMemoryContext()));
+            return lookupSourceProviderFuture;
         }
 
-        PartitioningSpillResult result = spiller.get().partitionAndSpill(page, spillMask);
-        spillInProgress = result.getSpillingFuture();
-        return result.getRetained();
-    }
-
-    public LocalPartitionGenerator getPartitionGenerator()
-    {
-        if (!partitionGenerator.isPresent()) {
-            partitionGenerator = Optional.of(new LocalPartitionGenerator(hashGenerator, lookupSourceFactory.partitions()));
+        private boolean tryFetchLookupSourceProvider()
+        {
+            if (lookupSourceProvider == null) {
+                if (!lookupSourceProviderFuture.isDone()) {
+                    return false;
+                }
+                lookupSourceProvider = requireNonNull(getDone(lookupSourceProviderFuture));
+                statisticsCounter.updateLookupSourcePositions(lookupSourceProvider.withLease(lookupSourceLease -> lookupSourceLease.getLookupSource().getJoinPositionCount()));
+            }
+            return true;
         }
-        return partitionGenerator.get();
-    }
 
-    @Override
-    public Page getOutput()
-    {
-        // TODO introduce explicit state (enum), like in HBO
+        private PartitioningSpillResult spillAndMaskSpilledPositions(Page page, IntPredicate spillMask)
```

Comment:
I think you still want to update `spillInProgress` here and validate that previous spill not failed.
This method is different than the previous one.